### ISSUE TITLE
Xgboost "predict_proba" method Killed with 21MM rows

### ISSUE THREAD

**Issue by aus10powell** [dmlc/xgboost#9907]
Thanks for participating in the XGBoost community! We use https://discuss.xgboost.ai for any general usage questions and discussions. The issue tracker is used for actionable items such as feature proposals discussion, roadmaps, and bug tracking.  You are always welcomed to post on the forum first :)

Issues that are inactive for a period of time may get closed. We adopt this policy so that we won't lose track of actionable issues that may fall at the bottom of the pile. Feel free to reopen a new one if you feel there is an additional problem that needs attention when an old one gets closed.

For bug reports, to help the developer act on the issues, please include a description of your environment, preferably a minimum script to reproduce the problem.

For feature proposals, list clear, small actionable items so we can track the progress of the change.

My setup:
I'm on a Kubernetes pod with 15Gb ram and 4000m cpu. The process is being killed when I call a trained XGBClassifier method "predict_proba(joined)" where joined is ~21MM rows. The memory just spikes and it's not that large. I'm able to run the exact same code on my local Mac machine so it must be a memory leakage issue. 

Is there a way to further debug this?

**Comment by trivialfis** [comment]
Does the machine have a sufficient amount of memory? If not, can you run the prediction in batches?

**Comment by aus10powell** [comment]
I don't have control over the library so I'd have to maintain on my own if I were to make that change. Also it's a 21,000,000 x 15 array of 32bit floats. It shouldn't overload the memory and doesnt on my local machine

**Comment by trivialfis** [comment]
I'm not sure how I could help given the information, could you please share:
- The XGBoost version you are using.
- The training parameters of the model.

**Comment by aus10powell** [comment]
- I updated xgboost to latest just to make sure that wasn't an issue (2.0.3)
- training parameters: n_estimators 20, learning rate 0.1 - 0.01, cv: 5
- numpy recarray (21000000x15) all float values
- Not GPU

**Comment by trivialfis** [comment]
So, how many classes there are? I'm trying to reproduce it using a generated dataset from `sklearn.datasets.make_classification`.

**Comment by aus10powell** [comment]
Thanks. There are only 2 classes. I'm monitoring the pod memory and it does spike over 100% even with a smaller amount of data around 1MMM. It seems like the memory overload happens fairly quickly. But again I can't reproduce outside of the Kubernetes Linux pod. In that case, I do receive the class probability.

**Comment by trivialfis** [comment]
Feel free to let me know if there's anything I can help with the k8s pod.

**Comment by aus10powell** [comment]
I'm just honestly stumped as to why the `xgboostTunedClassifier.estimator.predict_proba()` is apparently leaking memory on the linux platform and not on my local Mac machine. If you have any tips about how to trouble-shoot that would be great. I've tried memory profiling, garbage collection, all the usual suspects but nothing seems to be getting to the root of the matter.

**Comment by aus10powell** [comment]
Follow-up here: I have 32Gb locally on Mac that I used a different memory profiler for and looks like it is using the same amount of memory as the Kubernetes cluster. 

The main problem still remains though of why memory consumption 4x during the estimator.predict_proba step

**Comment by trivialfis** [comment]
If you are doing memory profiling, can the profiler tell where is the bottleneck?

**Comment by aus10powell** [comment]
No it doesn't seem to go into the internals

**Comment by aus10powell** [comment]
Error was conflict with internal library
